File: Backend/ride/test/AcceptRide.test.ts
import AcceptRide from '../src/application/usecase/AcceptRide';
import GetRide from '../src/application/usecase/GetRide';
import RequestRide from '../src/application/usecase/RequestRide';
import Signup from '../src/application/usecase/Signup';
import type DatabaseConnection from '../src/infra/database/DatabaseConnection';
import PgPromiseAdapter from '../src/infra/database/PgPromiseAdapter';
import LoggerConsole from '../src/infra/logger/LoggerConsole';
import AccountDAODatabase from '../src/infra/repository/AccountRepositoryDatabase';
import RideDAODatabase from '../src/infra/repository/RideRepositoryDatabase';

let signup: Signup;
let requestRide: RequestRide;
let getRide: GetRide;
let acceptRide: AcceptRide;
let databaseConnection: DatabaseConnection;

beforeEach(() => {
  databaseConnection = new PgPromiseAdapter();
  const accountDAO = new AccountDAODatabase(databaseConnection);
  const rideDAO = new RideDAODatabase();
  const logger = new LoggerConsole();
  signup = new Signup(accountDAO, logger);
  requestRide = new RequestRide(rideDAO, accountDAO, logger);
  getRide = new GetRide(rideDAO, logger);
  acceptRide = new AcceptRide(rideDAO, accountDAO);
});

test('Deve aceitar uma corrida', async function () {
  const inputSignupPassenger = {
    name: 'John Doe',
    email: `john.doe${Math.random()}@gmail.com`,
    cpf: '97456321558',
    isPassenger: true,
    password: '123456'
  };
  const outputSignupPassenger = await signup.execute(inputSignupPassenger);
  const inputRequestRide = {
    passengerId: outputSignupPassenger.accountId,
    fromLat: -27.584905257808835,
    fromLong: -48.545022195325124,
    toLat: -27.496887588317275,
    toLong: -48.522234807851476
  };
  const outputRequestRide = await requestRide.execute(inputRequestRide);
  const inputSignupDriver = {
    name: 'John Doe',
    email: `john.doe${Math.random()}@gmail.com`,
    cpf: '97456321558',
    carPlate: 'AAA9999',
    isDriver: true,
    password: '123456'
  };
  const outputSignupDriver = await signup.execute(inputSignupDriver);
  const inputAcceptRide = {
    rideId: outputRequestRide.rideId,
    driverId: outputSignupDriver.accountId
  };
  await acceptRide.execute(inputAcceptRide);
  const outputGetRide = await getRide.execute(outputRequestRide.rideId);
  expect(outputGetRide.status).toBe('accepted');
  expect(outputGetRide.driverId).toBe(outputSignupDriver.accountId);
});

test('Não pode aceitar uma corrida se a conta não for de um motorista', async function () {
  const inputSignupPassenger = {
    name: 'John Doe',
    email: `john.doe${Math.random()}@gmail.com`,
    cpf: '97456321558',
    isPassenger: true,
    password: '123456'
  };
  const outputSignupPassenger = await signup.execute(inputSignupPassenger);
  const inputRequestRide = {
    passengerId: outputSignupPassenger.accountId,
    fromLat: -27.584905257808835,
    fromLong: -48.545022195325124,
    toLat: -27.496887588317275,
    toLong: -48.522234807851476
  };
  const outputRequestRide = await requestRide.execute(inputRequestRide);
  const inputSignupDriver = {
    name: 'John Doe',
    email: `john.doe${Math.random()}@gmail.com`,
    cpf: '97456321558',
    isPassenger: true,
    password: '123456'
  };
  const outputSignupDriver = await signup.execute(inputSignupDriver);
  const inputAcceptRide = {
    rideId: outputRequestRide.rideId,
    driverId: outputSignupDriver.accountId
  };
  await expect(() => acceptRide.execute(inputAcceptRide)).rejects.toThrow(
    new Error('Only drivers can accept a ride')
  );
});

afterEach(async () => {
  await databaseConnection.close();
});

File: backend/ride-go/internal/application/usecase/accept_ride_test.go
package usecase

import (
	"database/sql"
	"errors"
	"testing"

	"github.com/google/uuid"
	"github.com/leonardograselalmeida/fake_uber/internal/domain/entity"
	"github.com/leonardograselalmeida/fake_uber/internal/infra/database"
	"github.com/leonardograselalmeida/fake_uber/internal/infra/repository"
	"github.com/stretchr/testify/assert"

	"github.com/DATA-DOG/go-sqlmock"
)

func Test_AcceptRide_When_Ride_And_Driver_Is_Valid(t *testing.T) {
	db, mock := database.NewMockDatabase()
	sut := makeSUT(db)
	passengerId := uuid.New()
	driver, _ := entity.CreateAccount("John Driver", "john.driver@gmail.com", "14181694046", "AAA9999", false, true)
	ride := entity.CreateRide(passengerId, 10.000, -10.000, 20.000, -20.000)

	acceptRideInput := AcceptRideInput{
		DriverId: driver.AccountId,
		RideId:   ride.RideId,
	}

	defer db.Close()
	accountRow := createAccountRow().
		AddRow(driver.AccountId, driver.Name, driver.Email, driver.Cpf, driver.CarPlate, driver.IsPassenger, driver.IsDriver)

	setQuerySelectToAccount(mock, driver.AccountId).
		WillReturnRows(accountRow)

	rideRow := createRideRow().
		AddRow(ride.RideId, ride.PassengerId, ride.DriverId, ride.Status, ride.FromLat, ride.FromLong, ride.ToLat, ride.ToLong, ride.Date)

	setQuerySelectToRide(mock, ride.RideId).
		WillReturnRows(rideRow)

	setQueryUpdateToRide(mock, entity.StatusAccept, driver.AccountId, ride.RideId).
		WillReturnResult(sqlmock.NewResult(0, 1))

	err := sut.Execute(acceptRideInput)

	if err != nil {
		t.Fatalf("Error in update ride: %s", err)
	}

	if err := mock.ExpectationsWereMet(); err != nil {
		t.Errorf("Error in querys: %s", err)
	}

	assert.NoError(t, err)
}

func Test_AcceptRide_When_Ride_Not_Exists_Sould_Return_Error(t *testing.T) {
	db, mock := database.NewMockDatabase()
	sut := makeSUT(db)
	passengerId := uuid.New()
	driver, _ := entity.CreateAccount("John Driver", "john.driver@gmail.com", "14181694046", "AAA9999", false, true)
	ride := entity.CreateRide(passengerId, 10.000, -10.000, 20.000, -20.000)

	acceptRideInput := AcceptRideInput{
		DriverId: driver.AccountId,
		RideId:   ride.RideId,
	}

	defer db.Close()
	accountRow := createAccountRow().
		AddRow(driver.AccountId, driver.Name, driver.Email, driver.Cpf, driver.CarPlate, driver.IsPassenger, driver.IsDriver)

	setQuerySelectToAccount(mock, driver.AccountId).
		WillReturnRows(accountRow)

	rideRow := createRideRow()
	setQuerySelectToRide(mock, ride.RideId).
		WillReturnRows(rideRow)

	err := sut.Execute(acceptRideInput)

	if err := mock.ExpectationsWereMet(); err != nil {
		t.Errorf("Error in querys: %s", err)
	}

	assert.EqualError(t, err, "ride not found")
}

func Test_AcceptRide_When_Driver_Not_Exists_Should_Return_Error(t *testing.T) {
	db, mock := database.NewMockDatabase()
	sut := makeSUT(db)
	passengerId := uuid.New()
	driver, _ := entity.CreateAccount("John Driver", "john.driver@gmail.com", "14181694046", "AAA9999", false, true)
	ride := entity.CreateRide(passengerId, 10.000, -10.000, 20.000, -20.000)

	acceptRideInput := AcceptRideInput{
		DriverId: driver.AccountId,
		RideId:   ride.RideId,
	}

	defer db.Close()
	accountRow := createAccountRow()

	setQuerySelectToAccount(mock, driver.AccountId).
		WillReturnRows(accountRow)

	err := sut.Execute(acceptRideInput)

	if err := mock.ExpectationsWereMet(); err != nil {
		t.Errorf("Error in querys: %s", err)
	}

	assert.EqualError(t, err, "driver not found")
}

func Test_AcceptRide_When_Get_Driver_Should_Return_Error_SqlQuery(t *testing.T) {
	db, mock := database.NewMockDatabase()
	sut := makeSUT(db)
	passengerId := uuid.New()
	driver, _ := entity.CreateAccount("John Driver", "john.driver@gmail.com", "14181694046", "AAA9999", false, true)
	ride := entity.CreateRide(passengerId, 10.000, -10.000, 20.000, -20.000)

	acceptRideInput := AcceptRideInput{
		DriverId: driver.AccountId,
		RideId:   ride.RideId,
	}

	defer db.Close()

	setQuerySelectToAccount(mock, driver.AccountId).
		WillReturnError(errors.New("erro ao consultar account no banco de dados"))

	err := sut.Execute(acceptRideInput)

	if err := mock.ExpectationsWereMet(); err != nil {
		t.Errorf("Error in querys: %s", err)
	}

	assert.EqualError(t, err, "erro ao consultar account no banco de dados")
}

func Test_AcceptRide_When_Account_Is_Passeger_Sould_Return_Error(t *testing.T) {
	db, mock := database.NewMockDatabase()
	sut := makeSUT(db)
	passengerId := uuid.New()
	driver, _ := entity.CreateAccount("John Driver", "john.driver@gmail.com", "14181694046", "", true, false)
	ride := entity.CreateRide(passengerId, 10.000, -10.000, 20.000, -20.000)

	acceptRideInput := AcceptRideInput{
		DriverId: driver.AccountId,
		RideId:   ride.RideId,
	}

	defer db.Close()
	accountRow := createAccountRow().
		AddRow(driver.AccountId, driver.Name, driver.Email, driver.Cpf, driver.CarPlate, driver.IsPassenger, driver.IsDriver)

	setQuerySelectToAccount(mock, driver.AccountId).
		WillReturnRows(accountRow)

	err := sut.Execute(acceptRideInput)

	if err := mock.ExpectationsWereMet(); err != nil {
		t.Errorf("Error in querys: %s", err)
	}

	assert.EqualError(t, err, "only drivers can accept a ride")
}

func Test_AcceptRide_When_Get_Ride_Should_Return_Error_SqlQuery(t *testing.T) {
	db, mock := database.NewMockDatabase()
	sut := makeSUT(db)
	passengerId := uuid.New()
	driver, _ := entity.CreateAccount("John Driver", "john.driver@gmail.com", "14181694046", "AAA9999", false, true)
	ride := entity.CreateRide(passengerId, 10.000, -10.000, 20.000, -20.000)

	acceptRideInput := AcceptRideInput{
		DriverId: driver.AccountId,
		RideId:   ride.RideId,
	}

	defer db.Close()
	accountRow := createAccountRow().
		AddRow(driver.AccountId, driver.Name, driver.Email, driver.Cpf, driver.CarPlate, driver.IsPassenger, driver.IsDriver)

	setQuerySelectToAccount(mock, driver.AccountId).
		WillReturnRows(accountRow)

	setQuerySelectToRide(mock, ride.RideId).
		WillReturnError(errors.New("erro ao consultar ride no banco de dados"))

	err := sut.Execute(acceptRideInput)

	if err := mock.ExpectationsWereMet(); err != nil {
		t.Errorf("Error in querys: %s", err)
	}

	assert.EqualError(t, err, "erro ao consultar ride no banco de dados")
}

func Test_AcceptRide_When_UpdateRide_Should_Return_Sql_Error(t *testing.T) {
	db, mock := database.NewMockDatabase()
	sut := makeSUT(db)
	passengerId := uuid.New()
	driver, _ := entity.CreateAccount("John Driver", "john.driver@gmail.com", "14181694046", "AAA9999", false, true)
	ride := entity.CreateRide(passengerId, 10.000, -10.000, 20.000, -20.000)

	acceptRideInput := AcceptRideInput{
		DriverId: driver.AccountId,
		RideId:   ride.RideId,
	}

	defer db.Close()
	accountRow := createAccountRow().
		AddRow(driver.AccountId, driver.Name, driver.Email, driver.Cpf, driver.CarPlate, driver.IsPassenger, driver.IsDriver)

	setQuerySelectToAccount(mock, driver.AccountId).
		WillReturnRows(accountRow)

	rideRow := createRideRow().
		AddRow(ride.RideId, ride.PassengerId, ride.DriverId, ride.Status, ride.FromLat, ride.FromLong, ride.ToLat, ride.ToLong, ride.Date)

	setQuerySelectToRide(mock, ride.RideId).
		WillReturnRows(rideRow)

	setQueryUpdateToRide(mock, entity.StatusAccept, driver.AccountId, ride.RideId).
		WillReturnError(errors.New("erro ao atualizar ride no banco de dados"))

	err := sut.Execute(acceptRideInput)

	if err := mock.ExpectationsWereMet(); err != nil {
		t.Errorf("Error in querys: %s", err)
	}

	assert.EqualError(t, err, "erro ao atualizar ride no banco de dados")
}

func makeSUT(db *sql.DB) AcceptRide {
	accountRepository := repository.AccountRepository{Db: db}
	rideRepository := repository.RideRepository{Db: db}

	sut := AcceptRide{AccountRepository: &accountRepository, RideRepository: &rideRepository}
	return sut
}

func createAccountRow() *sqlmock.Rows {
	return sqlmock.NewRows([]string{"account_id", "nome", "email", "cpf", "car_plate", "is_passenger", "is_driver"})
}

func createRideRow() *sqlmock.Rows {
	return sqlmock.NewRows([]string{"ride_id", "passenger_id", "driver_id", "status", "from_lat", "from_long", "to_lat", "to_long", "date"})
}

func setQuerySelectToAccount(mock sqlmock.Sqlmock, accountId uuid.UUID) *sqlmock.ExpectedQuery {
	return mock.ExpectQuery("select account_id, name, email, cpf, car_plate, is_passenger, is_driver from cccat14.account where account_id = $1").
		WithArgs(accountId)
}

func setQuerySelectToRide(mock sqlmock.Sqlmock, rideId uuid.UUID) *sqlmock.ExpectedQuery {
	return mock.ExpectQuery("select ride_id, passenger_id, driver_id, status, from_lat, from_long, to_lat, to_long, date FROM cccat14.ride where ride_id = $1").
		WithArgs(rideId)
}

func setQueryUpdateToRide(mock sqlmock.Sqlmock, status string, driverId uuid.UUID, rideId uuid.UUID) *sqlmock.ExpectedExec {
	return mock.ExpectExec("update cccat14.ride set status = $1, driver_id = $2 where ride_id = $3").
		WithArgs(status, driverId, rideId)
}
